primitives/src/merkle_tree/mod.rs

```diff
@@ -0,0 +1,288 @@
+// Copyright (c) 2022 Espresso Systems (espressosys.com)
+// This file is part of the Jellyfish library.
+
+// You should have received a copy of the MIT License
+// along with the Jellyfish library. If not, see <https://mit-license.org/>.
+
+//! Merkle Tree traits and implementations
+pub mod append_only;
+pub mod examples;
+pub mod macros;
+pub mod sparse_merkle_tree;
+
+mod internal;
+
+use crate::errors::PrimitivesError;
+use ark_std::{borrow::Borrow, fmt::Debug, string::ToString, vec, vec::Vec};
+use num_bigint::BigUint;
+use num_traits::ToPrimitive;
+use serde::{Deserialize, Serialize};
+
+#[derive(Clone, Copy, PartialEq, Eq, Debug, Hash)]
+/// The result of querying at an index in the tree
+/// Typically, F for element type, P for membership proof type and N for
+/// non-membership proof type
+pub enum LookupResult<F, P, N> {
+    /// The value at the given index, and a proof of validity
+    Ok(F, P),
+    /// The index is valid but we do not have the leaf in memory
+    NotInMemory,
+    /// The index is outside the occupied range in the tree, and a
+    /// non-membership proof
+    EmptyLeaf(N),
+}
+
+impl<F, P, N> LookupResult<F, P, N> {
+    /// Assert the lookup result is Ok. Return a tuple of element and membership
+    /// proof.
+    pub fn expect_ok(self) -> Result<(F, P), PrimitivesError> {
+        match self {
+            LookupResult::Ok(x, proof) => Ok((x, proof)),
+            LookupResult::NotInMemory => Err(PrimitivesError::InternalError(
+                "Expected Ok, found NotInMemory".to_string(),
+            )),
+            LookupResult::EmptyLeaf(_) => Err(PrimitivesError::InternalError(
+                "Expected Ok, found EmptyLeaf".to_string(),
+            )),
+        }
+    }
+
+    /// Assert the lookup result is NotInMemory. Return a non-membership proof.
+    pub fn expect_empty(self) -> Result<N, PrimitivesError> {
+        match self {
+            LookupResult::EmptyLeaf(n) => Ok(n),
+            LookupResult::Ok(..) => Err(PrimitivesError::InternalError(
+                "Expected EmptyLeaf, found element".to_string(),
+            )),
+            LookupResult::NotInMemory => Err(PrimitivesError::InternalError(
+                "Expected EmptyLeaf, found NotInMemory".to_string(),
+            )),
+        }
+    }
+}
+
+/// An element of a Merkle tree.
+pub trait Element: Clone + Eq + PartialEq {}
+impl<T: Clone + Eq + PartialEq> Element for T {}
+
+/// An index type of a leaf in a Merkle tree.
+pub trait Index: Debug + Eq + PartialEq + Ord + PartialOrd + Clone + ToTraversalPath {}
```

Comment:
`ToTraversalPath` seems like it would be used by the implementor of a Merkle tree, not the user. If true, it might be better to require that on the impl (e.g. `impl<I: ToTraversalPath> AppendableMerkleTreeScheme for MerkleTree<I>`) since the `ToTraversalPath` requirement might be stricter than a particular implementation really needs. (As a concrete example, `ToTraversalPath`, as currently specified, requires a conversion to a path for any arity, but any particular Merkle tree implementation probably has a fixed arity, see my comment below)